prover/src/constraints/evaluator.rs

```diff
@@ -23,18 +26,117 @@ use utils::{iterators::*, rayon};
+    type HashFn = H;
+
+    fn evaluate(
+        self,
+        trace: &Self::TraceLde,
+        domain: &'a StarkDomain<<E as FieldElement>::BaseField>,
+    ) -> ConstraintEvaluationTable<'a, E> {
+        assert_eq!(
+            trace.trace_len(),
+            domain.lde_domain_size(),
+            "extended trace length is not consistent with evaluation domain"
+        );
+
+        // build a list of constraint divisors; currently, all transition constraints have the same
+        // divisor which we put at the front of the list; boundary constraint divisors are appended
+        // after that
+        let mut divisors = vec![self.transition_constraints.divisor().clone()];
+        divisors.append(&mut self.boundary_constraints.get_divisors());
+
+        // allocate space for constraint evaluations; when we are in debug mode, we also allocate
+        // memory to hold all transition constraint evaluations (before they are merged into a
+        // single value) so that we can check their degrees later
+        #[cfg(not(debug_assertions))]
+        let mut evaluation_table = ConstraintEvaluationTable::<E>::new(domain, divisors);
+        #[cfg(debug_assertions)]
+        let mut evaluation_table =
+            ConstraintEvaluationTable::<E>::new(domain, divisors, &self.transition_constraints);
+
+        // when `concurrent` feature is enabled, break the evaluation table into multiple fragments
+        // to evaluate them into multiple threads; unless the constraint evaluation domain is small,
+        // then don't bother with concurrent evaluation
+
+        #[cfg(not(feature = "concurrent"))]
+        let num_fragments = 1;
+
+        #[cfg(feature = "concurrent")]
+        let num_fragments = if domain.ce_domain_size() >= MIN_CONCURRENT_DOMAIN_SIZE {
+            rayon::current_num_threads().next_power_of_two()
+        } else {
+            1
+        };
+
+        // evaluate constraints for each fragment; if the trace consist of multiple segments
+        // we evaluate constraints for all segments. otherwise, we evaluate constraints only
+        // for the main segment.
+        let mut fragments = evaluation_table.fragments(num_fragments);
+        iter_mut!(fragments).for_each(|fragment| {
+            if self.air.trace_info().is_multi_segment() {
+                self.evaluate_fragment_full(trace, domain, fragment);
+            } else {
+                self.evaluate_fragment_main(trace, domain, fragment);
+            }
+        });
+
+        // when in debug mode, make sure expected transition constraint degrees align with
+        // actual degrees we got during constraint evaluation
+        #[cfg(debug_assertions)]
+        evaluation_table.validate_transition_degrees();
+
+        evaluation_table
+    }
+}
+impl<
+        'a,
+        A: Air,
+        E: FieldElement<BaseField = A::BaseField>,
+        H: ElementHasher<BaseField = E::BaseField> + std::marker::Sync,
+    > DefaultConstraintEvaluator<'a, A, E, H>
+{
```

Comment:
Assuming we change the trait definition as descried above, we can change this to:
```Rust
impl<'a, A, E> DefaultConstraintEvaluator<'a, A, E>
where
    A: Air,
    E: FieldElement<BaseField = A::BaseField>,
{
```